generate-event.go

```diff
@@ -1,367 +1,158 @@
 package main
 
 import (
-	"encoding/base64"
-	"fmt"
-	"os"
-	"text/template"
-
+	"log"
 	"github.com/codegangsta/cli"
+	"strings"
+	"fmt"
+	"encoding/json"
+	"regexp"
 )
 
-var events = map[string]string{
-	"S3":       s3Event,
-	"SNS":      snsEvent,
-	"Kinesis":  kinesisEvent,
-	"DynamoDB": dynamodbEvent,
-	"Api":      apiEvent,
-	"Schedule": scheduleEvent,
+/**
+    The go:generate tag is used to generate a binary go file that
+    contains all data from the sample-events folder, which is used
+    by this file to generate commands for generate-event. If any changes
+    are made to the sample-events folder, the binary file will need to be
+    re-generated via running
+        'go generate'
+    in the project root folder. This command requires go-bindata, which can be
+    installed by running
+        'go get -u github.com/jteeuwen/go-bindata/...'
+ */
+// ***DO NOT EDIT THE FOLLOWING COMMENT - it is required for 'go generate' to work***
+//go:generate go-bindata -o generated-event-binary.go sample-events/...
+
+var nameToCommandMap map[string]string = make(map[string]string)
+var nameToDescriptionMap map[string]string = make(map[string]string)
+var nameToFlagListMap map[string][]Flag = make(map[string][]Flag)
+
+type CommandMap struct {
+	Commands []Command
 }
 
-func generate(eventType string, c *cli.Context) {
+type Command struct {
+	Name        string `json:"name"`
+	Command     string `json:"command"`
+	Description string `json:"description"`
+	Flags       []Flag `json:"flags"`
+}
 
-	sample, ok := events[eventType]
-	if !ok {
-		fmt.Printf("Unsupported event type: %s", eventType)
-		os.Exit(1)
-	}
+type Flag struct {
+	Name    string `json:"name"`
+	Usage   string `json:"usage"`
+	Value   string `json:"value"`
+}
 
-	t, err := template.New("event").Parse(sample)
+// Generates top-level commands for generate-event for all
+// categories under sample-events
+func generateEventCategories() (categories []cli.Command) {
+	mapFileToCommands()
+
+	dirs, err := AssetDir("sample-events")
 	if err != nil {
-		fmt.Printf("Failed to load sample %s event: %s", eventType, err)
-		os.Exit(1)
+		log.Fatal(err)
+	}
+
+	for _, dir := range dirs {
+		if (strings.Contains(dir, ".")) {
+			// Skip names that are not directories
+			continue
+		}
+		command := cli.Command{
+			Name:  strings.Replace(strings.ToLower(dir), " ", "-", -1),
+			Usage: "View available events for " + dir + "",
+			Subcommands: generateCategorySubCommands(dir),
+		}
+
+		categories = append(categories, command)
 	}
 
-	switch eventType {
-	case "S3":
+	return
+}
+
+func generateCategorySubCommands(subDir string) (subcommands []cli.Command){
+	path := "sample-events/" + subDir
+	files, err := AssetDir(path)
```

Comment:
`AssetDir` is a method from go-bindata right? You don't seem to include it here why?

Reply:
The function is from the generated binfile, and not from go-bindata, since go-bindata is just a tool and not a library

Reply:
Fair enough. So `generate-event-binary.go` is the file that provides these methods. It means ppl should `go generate` at least once before running `go build`, correct? In that case, is this section accurate? https://github.com/awslabs/aws-sam-local#build-from-source If not, can you update it?

Reply:
Yeah, I can update that. I do have a section under sample-events specifically for this, but I think it's probably better to mention it to the main README as well

https://github.com/stshan/aws-sam-local/tree/develop/sample-events

Reply:
sure thanks